examples/audio_decoder.rs

```diff
@@ -0,0 +1,70 @@
+extern crate servo_media;
+
+use servo_media::audio::buffer_source_node::AudioBufferSourceNodeMessage;
+use servo_media::audio::decoder::AudioDecoderCallbacks;
+use servo_media::audio::node::{AudioNodeMessage, AudioNodeType};
+use servo_media::ServoMedia;
+use std::env;
+use std::fs::File;
+use std::io::Read;
+use std::sync::mpsc;
+use std::sync::{Arc, Mutex};
+use std::{thread, time};
+
+fn run_example(servo_media: Arc<ServoMedia>) {
+    let mut graph = servo_media.create_audio_graph();
+    let args: Vec<_> = env::args().collect();
+    let filename: &str = if args.len() == 2 {
+        args[1].as_ref()
+    } else {
+        panic!("Usage: cargo run --bin audio_decoder <file_path>")
+    };
+    let mut file = File::open(filename).unwrap();
+    let mut bytes = vec![];
+    file.read_to_end(&mut bytes).unwrap();
+    let decoded_audio: Arc<Mutex<Vec<f32>>> = Arc::new(Mutex::new(Vec::new()));
+    let progress = decoded_audio.clone();
+    let (sender, receiver) = mpsc::channel();
+    let callbacks = AudioDecoderCallbacks::new()
+        .eos(move || {
+            sender.send(()).unwrap();
+        })
+        .error(|| {
+            eprintln!("Error decoding audio");
+        })
+        .progress(move |buffer| {
+            progress
+                .lock()
+                .unwrap()
+                .extend_from_slice((*buffer).as_ref());
+        })
+        .build();
+    graph.decode_audio_data(bytes.to_vec(), callbacks);
```

Comment:
Should we eventually make it such that this can be plugged into a source element on the render thread side, or do we want this to be separate to make the network stuff work?

Reply:
For media elements (and MediaElementAudioSourceNodes) we will likely use a different approach and use `GstPlayer` directly (which handles all the decoding work internally), check https://github.com/servo/media/issues/43.

For MediaStream*AudioSourceNodes, I am not so sure yet, I still need to read more of these specs, but in general, I have the feeling that all decoding should happen outside of any WebAudio source element, and certainly off the render thread.

In any case, `AudioDecoder` and `GStreamerAudioDecoder` are meant to be reusable.